### PR TITLE
switch to ReactDOM.createPortal

### PR DESCRIPTION
**Pull request by jochenberger**
Modal stops working in React 16 (#188). I thought this might fix it but it doesn't. It might be a good idea to do the switch anyway.

### CODE REVIEW

**Comment by jochenberger** [comment]
Just switching to `ReactDOM.createPortal` doen't help, it produces the same error.

File: src/Portal.js [hunk 1]
```diff
@@ -11,6 +11,9 @@ import ownerDocument from './utils/ownerDocument';
  * You can think of it as a declarative `appendChild()`, or jQuery's `$.fn.appendTo()`.
  * The children of `<Portal/>` component will be appended to the `container` specified.
  */
+
+const useCreatePortal = ReactDOM.createPortal !== undefined;
```
Comment (jochenberger) [inline src/Portal.js]: Should we also support the deprecated `ReactDOM.unstable_ createPortal`? IIRC, it's only present in React 16 pre-releases.

**Comment by jochenberger** [comment]
This does not fix (#188) but might be nice to have anyway.

**Comment by jquense** [comment]
Thanks KS for the help, you might want to looks at the tests in Rea t for how the new API works. It's not the same as the old one you actually call it during render()

**Comment by jochenberger** [comment]
I ran the tests with React 15 and 16 and they passed. But I can have another look.

**Comment by jquense** [comment]
I mean look at React's tests for createPortal,  not ours, for proper usage of the api :)

**Comment by jochenberger** [comment]
Getting closer. The tests that require `React.Children.only` to throw stop working, but since that also affects `PositionSpec`, I think it's an unrelated issue.
There are still two other tests failing in `PortalSpec` though.

**Comment by jochenberger** [comment]
I'm struggling with when to call `_mountOverlayTarget`. `_mountOverlayTarget` needs to find the DOM node for the container. It uses `ReactDOM.findDOMNode` via `utils/getContainer`, which means that it must not be called during render.

File: src/Portal.js (hunk 1 above)
Comment (taion) [reply src/Portal.js]: i'd go with "no"; just support the releases i think

**Comment by jochenberger** [comment]
`cDM` is too late because the overlay target needs to be initialized when we call `createPortal` from `render`.

File: src/Portal.js [hunk 2]
```diff
@@ -26,9 +29,21 @@ class Portal extends React.Component {
     ])
   };
 
+  componentWillMount(){
+    if (useCreatePortal){
```
Comment (jquense) [inline src/Portal.js]: this also needs a canUseDOM guard here, for serverside rendering. dom-helpers exports that constant for eas of use

File: src/Portal.js [hunk 3]
```diff
@@ -93,7 +113,12 @@ class Portal extends React.Component {
   }
 
   render() {
-    return null;
+    if (useCreatePortal && this.props.children) {
```
Comment (jquense) [inline src/Portal.js]: paired with my comment above, this needs to confirm `this._overlayTarget` exists for the SSR case

File: test/ModalSpec.js [hunk 4]
```diff
@@ -171,7 +171,7 @@ describe('Modal', function () {
       backdrop.style.borderWidth).to.equal('3px');
   });
 
-  it('Should throw with multiple children', function () {
+  xit('Should throw with multiple children', function () {
```
Comment (jquense) [inline test/ModalSpec.js]: why are these failing now?

Comment (jochenberger) [reply test/ModalSpec.js]: I have no idea. But the failure is not caused by my changes (`PositionSpec` fails too). I suspect, there are some changes around how errors are handled and/or reported in React 16.

**Comment by taion** [comment]
Given that this needs to mount its own extra portal element, I don't think you can use the cWM-based flow.

I think you need to create the element in cDM, then e.g. check `this.state` in `render` to see if we have a portal element.

**Comment by jquense** [comment]
So the new Portal API is awkward. We want to render the portal element on the first render _if possible_ to avoid missing updates or bubbling issues that the older api has. But yeah, it doesn't work if the mount container is isn't just a DOM node, or in the SSR case. I'm not sure what the best option forward is...I'm leaning towards just requiring a DOM node as a container vs a component instance but that might dramatically reduce the usefulness of the component?

Side note, the behavior between the old and new API is different, createPortal, allows events to bubble through as if it was a real child

**Comment by taion** [comment]
Wait why do we need `_overlayTarget` anyway here? There's no way we can safely `createElement` a `div` in cWM, is there?

**Comment by jquense** [comment]
You can safely create one if you check for `document` in a browser it's going to be there, its only in SSR env that you can't but that's gonna give the same result as if you double rendered and created in cDM

File: src/Portal.js (hunk 2 above)
Comment (jochenberger) [reply src/Portal.js]: done

File: src/Portal.js (hunk 3 above)
Comment (jochenberger) [reply src/Portal.js]: done

File: test/ModalSpec.js (hunk 4 above)
Comment (jochenberger) [reply test/ModalSpec.js]: I opened #207 to track that down.

**Comment by TryingToImprove** [comment]
Note that the examples does not work anymore because of `@monastic.panic/component-playground/Playground` which is still using `React.PropTypes`

**Comment by TryingToImprove** [comment]
The `.to.throw` is not working because the ` assertion-error` library is not included.

**Comment by jquense** [comment]
I updated `component-playground` to fix the propType issues. What do we think about this? Have we solved the "container that is a component instance" issue?

**Comment by jquense** [comment]
Ok, I'm realizing there is a bunch of code here (originally) that I don't really understand any more...

**Comment by jochenberger** [comment]
This is probably obsolete now that #208 is merged.

**Pull request closed without merging**